x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -839,6 +845,43 @@ private static LogicalPlan maybePushDownPastUnary(Filter filter, UnaryPlan unary
         }
     }
 
+    private record AttributeReplacement(List<Expression> rewrittenExpressions, AttributeMap<Alias> replacedAttributes) {};
+
+    /**
+     * Replace attributes in the given expressions by assigning them temporary names.
+     * Returns the rewritten expressions and a map with an alias for each replaced attribute; the rewritten expressions reference
+     * these aliases.
+     */
+    private static AttributeReplacement renameAttributesInExpressions(
+        Set<String> attributeNamesToRename,
+        List<? extends Expression> expressions
+    ) {
+        AttributeMap<Alias> aliasesForReplacedAttributes = new AttributeMap<>();
+        List<Expression> rewrittenExpressions = new ArrayList<>();
+
+        for (Expression expr : expressions) {
+            rewrittenExpressions.add(expr.transformUp(Attribute.class, attr -> {
+                if (attributeNamesToRename.contains(attr.name())) {
+                    Alias existingAlias = aliasesForReplacedAttributes.get(attr);
+                    if (existingAlias == null) {
+                        String tempName = SubstituteSurrogates.rawTemporaryName(attr.name(), "temp_name", (int) attr.id().toLong());
+
+                        Alias tempNameForShadowedAttr = new Alias(Source.EMPTY, tempName, attr);
```

Comment:
use aliasesForReplacedAttributes.computeIfMissing(attr, attr -> {}) which lazily creates the alias if it's missing.